docs/architecture/adr-066-keyring-plugin.md

```diff
@@ -0,0 +1,451 @@
+    // Client is defined above
+    // NewKeyringServiceClient is generated with protoc
+    return &Client{Client: NewKeyringServiceClient(c)}, nil
+}
+```
+
+#### Keyring Concrete Type
+```go
+import "github.com/hashicorp/go-plugin"
+
+// this is what will implement SDK keyring
+type PluginsKeyStore struct {
+	client  *plugin.Client
+	backEnd PluginKeyring // defined above
+	cdc     codec.Codec
+}
+
+func NewKeyring(...) *PluginsKeyStore {
+	client := plugin.NewClient(&plugin.ClientConfig{...})
+	// Connect via RPC
+	rpcClient, _ := client.Client()
+	
+	// Request the keyStore
+	raw, _ := rpcClient.Dispense("keyring") // here's where KeyringGRPC.GRPCClient() will be called
+
+	return &PluginsKeyStore{
+		client:  client,
+		backEnd: raw.(PluginKeyring), // PluginKeyring is defined above
+		cdc:     cdc,
+	}
+```
+
+#### Plugin Concrete type
+```go
+import "github.com/hashicorp/go-plugin"
+
+// backend is responsible for storing keys. It will have different flavors.
+type backend interface {
+	Get(string) (secret, error)
+	Set(string) error
+	Remove(string) error
+}
+
+type keyring struct {
+    cdc codec.Codec
+    db backend
+}
+
+func newKeyring(...) *keyring {...}
+
+func main() {
+	plugin.Serve(&plugin.ServeConfig{
+		HandshakeConfig: ...,
+		Plugins: map[string]plugin.Plugin{
+			"keyring": &KeyringGRPC{Impl: newKeyring(...)},
+		},
+		// A non-nil value here enables gRPC serving...
+		GRPCServer: plugin.DefaultGRPCServer,
+	})
+}
+
+```
+### Security
+
+Plugins must run in the same local network as the main process thus some security is guaranteed as the data never
+leaves the same machine. In any case connection can be secured with tls. 
+
+### Project Structure
+
+This new keyring could be implemented either in the keyring package of the SDK or in a new repository.
+
+#### New Repository
+If it is implemented in a different repository, the `keyring.New` method should be updated to support
+the instantiation of a keyring for plugins.
+
+A new dependency will be added to the SDK.
+
+#### Keyring Package
+If the new keyring implementation is defined in the keyring package, it would be beneficial to
+consider the following cosmetic refactors:
+ * keyring.go: This file should define abstract types only, along with some common functions.
+ * keyStore: Move the [keyStore](https://github.com/cosmos/cosmos-sdk/blob/v0.47.3/crypto/keyring/keyring.go#L206) code to its own file, along with all its associated functions. 
+ * factory.go: Create a factory file responsible for instantiating different keyrings. 
+
+
+## Consequences
+
+### Backwards Compatibility
+
+Backwards compatibility is guaranteed as the current keyring implementation and this new one can coexist.
+
+### Positive
+
+As plugins communicate with the main process using gRPC, plugins can be written in any language.
+
+Teams can easily develop plugins to meet their specific requirements,
+which opens the door to new functionalities in key management.
+
+### Negative
```

Comment:
What is the overhead compared to the current implementation? In terms of latency.

Reply:
The methods affected by this are those related to key creation and retrieval. Ultimately, it adds one more call to the codec. I'm not sure about the exact latency impact.